Enhance OpenGL Renderer with Planet Rendering

raniabbrk opened the pull request:
#### **Description**

This pull request introduces significant enhancements to the rendering pipeline, improving modularity and maintainability. Key changes include refactoring the `Renderer` class, introducing a dedicated `Planet` class, and leveraging reusable utilities for spherical geometry generation. These updates ensure consistency across the rendering of skybox and planet objects, adhering to clean architecture principles.

#### **Changes Introduced**

1. **Planet Class**:
   - Introduced the `Planet` class for rendering planets with spherical geometry.
   - Implemented texture application and scaling for realistic planet visualization.
   - Inherits from the `Object` class, aligning with the established architecture.

2. **SphericalRenderer Class**:
   - Ensured encapsulation and improved method accessibility to support the new architecture.

3. **Renderer Class**:
   - Integrated planet rendering alongside the skybox.
   - Simplified rendering responsibilities for clear separation of concerns.

4. **SkyBox Class**:
   - Optimized geometry generation by using the shared `GeometryUtils` utility.
   - Ensured consistent rendering behavior and alignment with the updated architecture.

5. **GeometryUtils :**
   - Provides reusable utilities for generating spherical geometry, ensuring consistency across `SkyBox` and `Planet`.


---

#### **Benefits**
- **Modularity**: Clear separation of concerns across classes and components.
- **Reusability**: Shared utilities and components streamline rendering logic for spherical objects.
- **Maintainability**: Simplified architecture makes it easier to extend or debug the rendering pipeline.
- **Consistency**: Unified handling of transformations and geometry generation across the scene.

--- 

This pull request improves the rendering architecture significantly while ensuring future scalability.

raniabbrk committed:
feat: Add utility for generating spherical geometry in OpenGL

This commit introduces the `Sphere` utility object, which provides methods for generating
tessellated spherical geometry for OpenGL rendering.

raniabbrk committed:
feat: Add OpenGL-based Planet rendering class with configurable geometry and textures

This commit introduces the `Planet` class for rendering 3D planets in an OpenGL environment. Key features include:
- Sphere tessellation for realistic spherical geometry.
- Customizable texture support for planet surfaces.
- Configurable position, scaling, and appearance.
- Integration with OpenGL shaders for efficient rendering.

The class provides a framework for adding planets to 3D environments, supporting dynamic scaling and positioning.

raniabbrk committed:
refactor: Update SkyBox implementation to use utility-based spherical geometry generation

This refactor replaces the custom geometry generation logic in the `SkyBox` class with a call to the `generateSphericalGeometry` utility method. Key improvements include:

- Simplified and modularized geometry generation using reusable code.
- Improved maintainability and consistency across components by reusing `generateSphericalGeometry`.
- Ensures alignment with the approach used in other renderable objects, such as the `Planet`.

The refactor preserves all existing functionality, including rendering and texture mapping for the skybox.

raniabbrk committed:
feat: Add planet rendering to OpenGL renderer alongside skybox

This update enhances the `Renderer` class by introducing planet rendering capabilities. Key changes include:

- Integration of the `Planet` class for rendering 3D planets within the OpenGL environment.
- Initialization and configuration of planet geometry, textures, and rendering pipeline.
- Modification of the rendering loop to include the planet, ensuring proper depth handling alongside the skybox.
- Adjusted camera view matrix setup for a better 3D perspective.

The renderer now supports both a dynamic skybox and planets, providing a more comprehensive scene rendering experience.

raniabbrk committed:
feat: Add planet texture image for rendering

This commit adds a texture image to be used for rendering planets in the OpenGL environment.

AdrienBousquieEPFL commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/Renderer.kt:
The renderer shouldn’t contain any projection matrices, as all of this is already managed in the `Camera`, particularly for the objects. This will help keep responsibilities clear and avoid duplication.
Reference:
- #159

AdrienBousquieEPFL commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/renderables/Planet.kt:
```diff
@@ -0,0 +1,178 @@
+ * @property stepsPerBand The number of longitude steps per latitude band. Higher values improve
+ *   rendering fidelity.
+ */
+class Planet(
+    private val context: Context,
+    private val name: String = "Planet",
+    private val position: FloatArray = floatArrayOf(0.0f, 0.0f, 0.0f),
+    private val textureId: Int,
+    private val numBands: Int = DEFAULT_NUM_BANDS,
+    private val stepsPerBand: Int = DEFAULT_STEPS_PER_BAND
+) {
```
To maintain consistency with our structure, please have this class inherit from the `Object` class. Feel free to modify the `Object` class if you think it can be improved, but ensure that it aligns with the current architecture. Let me know if you need any input!

AdrienBousquieEPFL commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/renderables/utils/GeometryUtils.kt:
```diff
@@ -0,0 +1,102 @@
+ * The sphere is tessellated into latitude bands and longitude steps, which control the level of
+ * detail and smoothness of the sphere.
+ */
+object Sphere {
```
Would it make sense to rename this object to `GeometryUtils`? It might better reflect its purpose and align with our naming conventions.

AdrienBousquieEPFL commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/renderables/utils/GeometryUtils.kt:
```diff
@@ -0,0 +1,102 @@
+object Sphere {
+  fun generateSphericalGeometry(numBands: Int = 20, stepsPerBand: Int = 28): GeometryData {
+    val vertices = mutableListOf<Float>()
+    val colors = mutableListOf<Int>()
```
I’m not entirely sure, but I think we might not need `color` anymore since we’re now consistently using a texture. It might be worth considering whether we can remove this to simplify the code. What do you think?

raniabbrk replied:
You're right, thank you for noticing!

AdrienBousquieEPFL commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/renderables/Planet.kt:
```diff
@@ -0,0 +1,178 @@
+   *
+   * @param mvpMatrix A 4x4 matrix that combines the model, view, and projection transformations.
+   */
+  fun render(mvpMatrix: FloatArray) {
```
I think it makes sense to rename this to `draw` and use the `Camera` instead of a precomputed MVP matrix. It will help keep the terminology and functionality consistent with the rest of the implementation.

AdrienBousquieEPFL commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/renderables/Planet.kt:
```diff
@@ -0,0 +1,178 @@
+   * Initializes the planet by generating its geometry, setting up vertex buffers, and loading the
+   * texture. This method prepares the planet for rendering.
+   */
+  fun initialize() {
```
It would make sense to move this into an `init {}` block.

AdrienBousquieEPFL reviewed (dismissed):
The code implements what is described in the issue, though I’m currently unable to test the implementation and will do so later. However, the code doesn’t fully adhere to some rules we established in #159 (see the reviews) and should be updated to align with those guidelines. 

That said, don’t hesitate to propose ideas about how the current architecture could be adjusted to better fit your implementation. I’m open to discussing changes—just let me know before making any major updates so we can ensure the new guidelines are consistently applied across the codebase.

raniabbrk committed:
feat: Add SphereRenderer base class to eliminate duplicate code in sphere rendering.

- Created `SphereRenderer` as an abstract base class for reusable spherical geometry rendering.
- Consolidated buffer initialization and shader management into shared methods.
- Supports customizable tessellation with `numBands` and `stepsPerBand`.
- Simplifies derived classes such as `Planet` and `SkyBox` by centralizing common functionality.

raniabbrk committed:
refactor: change Planet class to extend SphereRenderer and enhance rendering workflow.

- Refactored `Planet` to inherit from `SphereRenderer` for reusable geometry handling.
- Simplified buffer and shader initialization using `initializeBuffers` and `initializeShaders`.
- Maintained existing functionality while improving code modularity and reuse.

raniabbrk committed:
refactor: change SkyBox class to extend SphereRenderer and streamline functionality.

- Refactored the `SkyBox` class to inherit from `SphereRenderer` for reusable geometry handling.
- Simplified buffer and shader initialization using `initializeBuffers` and `initializeShaders`.
- Added detailed JavaDoc comments to all methods, fields, and the class.
- Ensured backward compatibility with the previous functionality while improving modularity.

raniabbrk committed:
refactor: Improve Renderer class modularized code.

- Improved clarity in method purpose and responsibilities.
- Enhanced initialization logic with modularized `intializeObjects` and `drawObjects`.
- Maintained consistency in naming and comments.

raniabbrk committed:
refactor(planet): Move initialization logic into init block and rename render to draw

- Moved initialization logic into the `init` block for better organization.
- Renamed the `render` method to `draw` and refactored it to use the `Camera` directly instead of a precomputed MVP matrix.
- Ensured consistency with the `Object` class structure by making `Planet` inherit from `Object`.

raniabbrk committed:
refactor(utils): Rename Sphere to GeometryUtils and remove color handling

- Renamed the `Sphere` object to `GeometryUtils` to better reflect its purpose and align with naming conventions.
- Removed support for color attributes in spherical geometry generation, as the implementation now consistently uses textures.
- Simplified the `generateSphericalGeometry` method by focusing only on essential attributes (vertices, indices, and texture coordinates).
- Improved clarity and reduced complexity in the codebase.

raniabbrk committed:
fix(renderer): Remove projection matrix handling and rely on Camera for object transformations

- Removed redundant projection matrix handling from the `Renderer` class.
- Simplified rendering responsibilities for better separation of concerns.

raniabbrk committed:
refactor(spherical-renderer): Make SphericalRenderer open and decouple from Planet inheritance

- Changed `SphericalRenderer` from abstract to open to allow flexible instantiation without requiring inheritance.
- Decoupled `SphericalRenderer` from `Planet`, as `Planet` now inherits from `Object` for consistent architecture.
- Updated access modifiers to ensure encapsulation while supporting composition-based usage.
- Improved overall flexibility and maintainability of the rendering logic.

Kenzoud commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/Renderer.kt:
```diff
@@ -17,50 +19,93 @@ class Renderer(private val context: Context) : GLSurfaceView.Renderer {
+  private val mvpMatrix = FloatArray(16)
+  private val viewMatrix = FloatArray(16)
+  private val projectionMatrix = FloatArray(16)
+
```
As Adrien said, all of the matrices matter is already handled in the Camera class. So this shouldn't be here.

Kenzoud commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/Renderer.kt:
```diff
@@ -17,50 +19,93 @@ class Renderer(private val context: Context) : GLSurfaceView.Renderer {
+
+    // Initialize camera matrices
+    Matrix.setIdentityM(camera.viewMatrix, 0)
+    Matrix.setIdentityM(projectionMatrix, 0)
```
These matrices are already initialized in the Camera class, so this can also go.

Kenzoud commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/Renderer.kt:
```diff
@@ -17,50 +19,93 @@ class Renderer(private val context: Context) : GLSurfaceView.Renderer {
+        0f,
+        1f,
+        0f // Up vector
+        )
```
There is no need to have this, this has been discussed in PR #159 :)

Kenzoud commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/renderables/utils/GeometryUtils.kt:
```diff
@@ -0,0 +1,94 @@
+
+    // Generate vertices
+    for (band in 0 until numBands) {
+      val intensity =
```
This val is unused, could be removed.

Kenzoud reviewed (changes requested):
The effort to modularize and generalize parts of the code, such as the shared utilities for geometry generation, is a great step toward reducing code repetition and ensuring consistency, good job! However, there are a few areas where further adjustments are needed to align better with existing patterns (PRs #159 #186) and ensure coherence with the previously implemented rendering pipeline.

raniabbrk committed:
refactor(geometry-utils): Remove unused intensity calculation to simplify geometry generation

raniabbrk committed:
fix(renderer): Delegate all matrix handling to Camera for cleaner responsibility

Kenzoud reviewed (dismissed):
Thank you for integrating the changes proposed earlier. However, after further review of the code I have noticed a major issue that should be addressed. Right now, the planet remains fixed in the center of the screen regardless of camera movement which means that the rendering logic currently ties the camera's focus or position to the planet's position. Instead of the planet being a fixed object in space, it behaves as though the camera is "orbiting" it or it's always the center of the scene.

The issue likely resides in the Planet's draw method. Where we might need proper logic to determine whether the camera is looking at the planet or elsewhere. Instead of applying a generic transformation, you maybe need to account for the camera's position and orientation to ensure the planet behaves like a fixed object in the world.

The planned PR #181  (Implement stars and planet display position logic) will introduce actual positions for celestial objects, which will further refine the behavior. However, the issue where the camera is "locked" to the planet needs to be resolved before that for a functional baseline. This could potentially be done in another PR if it seems too heavy. What do you think?

raniabbrk committed:
fix(rendering): correct texture rendering on SphereRenderer

- Fixed texture coordinate binding in `bindShaderAttributes` to ensure proper UV mapping.
- Adjusted shader attribute setup for `vTexCoord` to prevent texture distortion.
- Updated depth testing settings to avoid rendering artifacts during sphere rendering.
- Verified and corrected texture wrapping and filtering parameters.

This resolves issues with incorrect texture display and ensures smooth rendering of planetary surfaces.

raniabbrk committed:
refactor(rendering): update transformation pipeline and camera handling for Planet class

- Refactored `draw` method in `Planet` to correctly apply Model-View-Projection matrix calculations.
- Adjusted the planet's default position to be at `z = -2.0` for improved visibility within the scene.
- Ensured transformations (translation, scaling, projection) are applied in the correct order:
  `MVP = Projection * View * Model`.
- Simplified matrix initialization and removed redundant operations in the rendering process.
- Retained texture binding and shader attribute setup for consistent texture rendering.

These changes improve code clarity and ensure that the planet is correctly rendered in the 3D space relative to the camera.

rihabbelmekki reviewed (changes requested):
Overall, the implementation is well-structured and follows good OpenGL rendering practices. 

During the review, I noticed that, in your draw method, you're setting up new matrix arrays (viewMatrix, projMatrix, mvpMatrix) in each draw call instead of reusing existing ones. Consider copying the camera matrices in your draw method in your Planet.kt file. When I tested the implementation, this approach helped ensure more accurate matrix transformations and reduced unnecessary object allocations. 

**Proposed change** (to implement in the draw method):
```
// Copy camera matrices to ensure accurate positioning
System.arraycopy(camera.viewMatrix, 0, viewMatrix, 0, 16)
System.arraycopy(camera.projMatrix, 0, projMatrix, 0, 16)
```

raniabbrk committed:
chore: resolve merge conflicts

AdrienBousquieEPFL commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/Renderer.kt:
```diff
@@ -92,31 +99,54 @@ class Renderer : GLSurfaceView.Renderer {
 
-    val ratio: Float = width.toFloat() / height.toFloat()
+  /** Initializes additional objects in the scene. Currently includes a planet. */
+  private fun intializeObjects() {
+    planet = Planet(context, textureId = R.drawable.planet_texture) // Create planet
+  }
 
-    camera.updateProjectionMatrix(ratio)
+  /** Renders additional objects in the scene using the current MVP matrix. */
+  private fun drawObjects() {
+    planet.draw(camera) // Render the planet
   }
```
Does this method add significant value? It seems like it might not be necessary. What do you think?

raniabbrk replied:
I think it's a method that will help structure the code better once we will implement all types of objects (stars, constellations, etc.)

AdrienBousquieEPFL commented on app/src/main/java/com/github/lookupgroup27/lookup/model/map/Renderer.kt:
Thank you for adding the documentation—it really helps in better understanding the OpenGL rendering system!

raniabbrk replied:
You're welcome!

AdrienBousquieEPFL reviewed (commented):
Here’s my partial review for this PR. Great job overall—it’s impressive how much progress has been made in just one week of development. This will definitely help us move forward!

rihabbelmekki reviewed (approved):
Overall, the implementation is good. The changes improve the modularity and maintainability of the rendering pipeline, especially with the addition of the Planet class, SphericalRenderable and GeometryUtils. The refactor seems well-structured and aligns with clean architecture principles. Approved!
